Dapp-wallet communication standard (#27)
massalabs/massa-standards: wallet/dapps-communication.md

## Targeted Audience
This standard is targeted towards developers who are building decentralized applications (DApps) and cryptocurrency wallets that need to interact with each other in a browser environment. 

The standard assumes a working knowledge of web development and blockchain technology. It is also assumed that the reader has experience with browser extensions and understands the basic principles of secure communication between web pages and extensions.

> _NOTE:_ For more information on [content scripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts) and [background scripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts), see [MDN's browser extensions pages](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions).

## Specification

### Event-Based Communication

> Communication between the DApp and the wallet will use event-based messaging. There are two types of events: 

Author: we have a misunderstanding here... let's discuss about it ASAP

Author: it's okay in fact, the example implementation bellow follow what is explain here.

- those used by the extension to communicate with the web page. Those events are triggered on a static target: `window.massaWalletProvider`;

> - those used by the web page to communicate with the extension. Those events are triggered on a extension specific target: `massaWalletProvider-<wallet provider name>`.

Author: to be consistent with the previous bullet point, we should say windows.[...]

> _NOTE:_ If the wallet provider is named AwesomeWallet, the target for this extension would be: `massaWalletProvider-AwesomeWallet`.

### Commands

#### Register

This event is used by the extension to register itself to the webpage as a wallet provider.

| Direction            | Type       | Format                                      | Example                                                         |
| -------------------- | ---------- | -------------------------------------------- | --------------------------------------------------------------- |

> | Extension to webpage | `register` | <pre><code class="highlight-source-yaml">type: object<br>properties:<br>  id:<br>    type: string<br>  name:<br>    type: string<br>required:<br>  - id<br>  - name</code></pre>  | <pre><code class="highlight-source-json">{<br>  "id": "awesomeWalletprovider",<br>  "name": "Your Awesome Wallet Provider"<br>}</code></pre> |

Author: the markdown rendering is not as expected 
![image](https://user-images.githubusercontent.com/11590067/225037775-65af9d2b-6fa6-4b18-a763-d0b0655f62ec.png)

Reviewer: :sob:

| Webpage to extension | `account.balance` | <pre><code class="highlight-source-yaml">type: object<br>properties:<br>  address:<br>  type: string<br>required:<br>  - address</code></pre>  | <pre><code class="highlight-source-json">{<br>  "address": "A12..."<br>}</code></pre> |
| Extension to webpage | `account.balance.response` | <pre><code class="highlight-source-yaml">type: object<br>properties:<br>  balance:<br>    type: string<br>    format: bigInt<br>required:<br>  - balance</code></pre> | <pre><code class="highlight-source-json">{<br>  "balance": 1000n<br>}</code></pre> |

##### Delete

| Direction            | Type       | Format                                      | Example                                                         |
| -------------------- | ---------- | -------------------------------------------- | --------------------------------------------------------------- |
| Webpage to extension | `account.delete` | <pre><code class="highlight-source-yaml">type: object<br>properties:<br>  address:<br>    type: string<br>required:<br>  - address</code></pre>  | <pre><code class="highlight-source-json">{<br>  "address": "A12..."<br>}</code></pre> |

> | Extension to webpage | `account.delete.response` | <pre><code class="highlight-source-yaml">type: object<br>properties:<br>  response:<br>    type: string<br>    enum: ["OK", "REFUSED", "ERROR"]<br>  error:<br>    type: string<br>required:<br>  - response</code></pre> | <pre><code class="highlight-source-json">{<br>  "response": "REFUSED"<br>}</code></pre> |

Author: error property is missing

Reviewer: This property is optional (not required).

<td>

```yaml
type: object
properties:
  address:
  type: string
required:
  - address
```

</td>
<td>

```json
{
  "address": "A12..."
}
```

</td>
</tr>
<tr>
<td>Extension to webpage</td>
<td><code>account.delete.response</code></td>
<td>

```yaml
type: object
properties:
  response:
    type: string
    enum: ["OK", "REFUSED", "ERROR"]
  error:
    type: string
required:
  - response
```

</td><td>
  
```json
{
  "response": "REFUSED"
}
```

</td></tr></tbody></table>

##### Import

<table>
<thead>
<tr>
<th>Direction</th>
<th>Type</th>
<th>Format</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr>
<td>Webpage to extension</td>
<td><code>account.import</code></td>
<td>

```yaml
type: object
properties:
  privateKey:
    type: string
  publicKey:
    type: string
required:
  - privateKey
  - publicKey
```

</td><td>

```json
{
  "privateKey": "S12...",
  "publicKey": "P12...",
}
```

</td>
</tr>
<tr>
<td>Extension to webpage</td>
<td><code>account.import.response</code></td>
<td>

```yaml
type: object
properties:
  response:
    type: string
    enum: ["OK", "REFUSED", "ERROR"]
  error:
    type: string
required:
  - response
```

</td><td>
  
```json

> {
>   "response": "ERROR",
>   "error": "No connection with blockchain"
> }

Reviewer 2: Non blocking, of course: isn't better use **message** as key instead of **error** since **REFUSED** could also be followed by a message to inform the reason was refused!? 🤔  (not sure if is applicable).

```suggestion
{
  "response": "ERROR",
  "message": "No connection with blockchain"
}
```

Reviewer 2: Non blocking: would we possible run in a situation where we want to inform DEVs (on browser console) about something deprecated!? A new field missing, a new way to sign, a new version, etc!?

If yes, shouldn't we answer back using the well know HTTP standards? 200 (ok), 403 (forbidden), 426 (upgrade require) , etc? 🤔

Reviewer: > Non blocking, of course: isn't better use **message** as key instead of **error** since **REFUSED** could also be followed by a message to inform the reason was refused!? thinking (not sure if is applicable).

Cool, we should do that

Reviewer: > Non blocking: would we possible run in a situation where we want to inform DEVs (on browser console) about something deprecated!? A new field missing, a new way to sign, a new version, etc!?
> 
> If yes, shouldn't we answer back using the well know HTTP standards? 200 (ok), 403 (forbidden), 426 (upgrade require) , etc? thinking

Don't you think that the message is sufficient and easier to understand than HTTP response code for a human (I don't believe that this kind of returned will be automaticaly processed by the code...) ?

Reviewer 2: Indeed, it could be overkill. I don't disagree... I was just considering that potentially we could have more options than the 3 ones: **ok**, **refused**, **error**. If we adopt the well know HTTP palette we have intrinsic all kind of standard possibilities (if needed, of course). 

However, in the future, by some reason, we decide to include a new status (WARNING) for example, people need to be informed about the new status to eventually rely on that answer. Instead, using HTTP they are already aware that almost all range of API answer are covered.

But, please, feel free to go with you have defined. It's completely acceptable and simple.

Author: I have renamed error into message.

I think the enum is enough and http status code is too much.

Reviewer 2: Thank you! 

Sure, no worries for the status codes.

## Specification

Here is pseudo code for the two points of view:

- browser extension developer
- dapp builder

**As a browser extension developer, I would write...**

browser extension code, [content script](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts):

```text
// the extension register itself
if (window.massaWalletProvider == undefined) { /* wait */ }

window.bearbyWalletProvider = new EventTarget()

window.massaWalletProvider.addEventListener('loaded', () => {
  window.massaWalletProvider.dispatchEvent(new CustomEvent('register', {
    ProviderName: "bearby",
    eventTarget: "bearbyWalletProvider",

>     extensionID: "" // maybe not needed

Author: extensionID is intended to be used for communication between content_script and background_script, but we need to check the default behavior to determine if it's really needed.

      break
});
```

**Code snippet of massa javascript (or typescript) library:**

```text
registeredProviders = []
actions = [] // all the ongoing actions (sign, getBalance...)

// we start by listening register event
window.massaWalletProvider = new EventTarget()
window.massaWalletProvider.addEventListener('register', (payload) => {
  // add the new provider to the list
  // add the attribute wallets: list of wallets of this new provider
  // the first wallet is a newly created wallet, we give as constructor argument the event target name created by the extension
  registeredProviders.push({...payload, wallets: [new Wallet(payload.eventTarget)])
})

window.massaWalletProvider.dispatchEvent('loaded') // all wallet will catch this event and emit register event

export class Wallet {
  construct(private eventTargetName) {}

  function sign(payload) {
    const action = {
      correlationId: uuid() // pick a correlation id to track the sign message request, it can simply be a int incremented
      payload,
      response: null
    }
    actions.push(action)

    // listen for the response

>     window[eventTargetName].addEventListener('signed', ({ pkey, signature, correlationId }) => {

Author: the problem here is that we add a listener at each Wallet.sign, we must either add only one listener for the life of the wallet (for example in the constructor), either remove the listener when we have the response (for this we can use parameter `once`).

      actions.find(a => a.correlationId == correlationId).response = { pkey, signature }
    })

    // trigger the signature

>     window[eventTargetName].dispatchEvent(new CustomEvent('sign', action))

Author: we can here make a cancelable event https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#add_an_abortable_listener

    return new Promise((resole, reject) => {
      // reject in 60 seconds to not let the user wait for hours
      setTimeOut(() => {
        reject('event timeout')
      }, 60000)

      setInterval(() => {
        if (action.response != null) {
          resolve(action.response) // here we define what Wallet.sign will return when doing await on it
        } 
      }, 500)
    })
  }
}

```

**As a dapp builder, I would write...**

```text
providers: [] = await massa-js-library.listWalletProviders()

for each providers as p
  for each p.wallets as wallet
    print wallet.getAddress()

>     print wallet.getBalance()

Author: those 2 methods will return promise

wallet = providers.listWallets()[0]
pkey, signature = await wallet.sign([0, ...])
.then(response => {
  { pkey, signature } = response
  console.log(pkey)
  console.log(signature)

> })

Author: you can also .catch this promise, to display a message to the user when the 60 seconds timeout is fired.

```

> **Code snippet of massa javascript (or typescript) library:**

Author: I don't know if it's a good idea to split massa library code. we've already declared a peace of code above

## Specification

Here is pseudo code for the two points of view:

- browser extension developer
- dapp builder

**As a browser extension developer, I would write...**

browser extension code:

```text
// the extension register itself

> window.massaWalletProvider.addEventListener('loaded', () => {

Reviewer: I'm not sure it will work like that.
The first thing to do I guess is to create an html element to listen event on.

Then, you can trigger the register event.

Reviewer 3: I suggest something we have already done before in the massa-wallet with Antoine and it has worked really well (see here: https://github.com/massalabs/massa-standards/pull/26/files#diff-dad455c0814fa2229eaae244a61b26f2f312670c0d14b6da452cd46744739956R66-R93) and https://github.com/massalabs/massa-wallet/blob/main/inject.js

It is similar to what Nathan has, but it is a more advanced mechanism and is closer to what Metamask has done.

Reviewer: I don't see why it is more advanced, can you clarify this part ?
From my point of view, using events on `window` element is a bad practice, creating potential conflict with other extension. I would prefer to use dedicated html elements.

Can you complete this section based on my remarks and what you suggest?

Reviewer 3: @gregLibert  I dont understand what you mean by dedicated html elements ? An html element is a script, isn't it ? Since I dont understand your point here, maybe you can elaborate and expand your idea

Reviewer: Yes, sorry.

Back when I was developing browser entensions more than 10 years ago, I used to do this in an html element (object under window.document), but it seems that now you can [create](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/EventTarget) an [EventTarget] (https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) wherever you want.

My point is that adding an event listner and firing an event directly on the window element is a bad practice  because if it is done correctly, there is no conflict, but if the page dev doesn't do it right, there can be side effects that are hard to debug.
I would prefer that we use a dedicated EventTarget. @Thykof suggests using the JS object `window.massaWalletProvider`, which is fine with me.

Reviewer 4: More like this ?
```
// Wallet Side
new EventTarget().addEventListener('loaded', function(event) {
  console.log('Web3 Loaded', event.detail);
});

// Dapp Side
const myEvent = new CustomEvent('loaded', {
  detail: {
    message: 'You can register'
  }
});
myEventTarget.dispatchEvent(myEvent);
```

Author: so the html element to listen on in fact not a html element but an EventTarget. I have created one in the massa-library side that handle only registration of wallet providers.

Then each wallet provider (extension) define their own EventTarget

  window.massaWalletProvider.emit('register', {
    ProviderName: "bearby",
    htmlObject: "window.bearyWalletProvider"
  })
})
```

> 

Reviewer: Here is missing the logic for the bearby wallet to:
- return wallets (listWallets function)
- for a given wallet, return Address, balance and sign command

Reviewer 3: The wallet functionality I presume needs to be concealed inside the injected js into the dom (I have suggested this here: https://github.com/massalabs/massa-standards/pull/26/files#diff-dad455c0814fa2229eaae244a61b26f2f312670c0d14b6da452cd46744739956R73). It contains all the funcs we need the interface to implement.

Reviewer: No, my question is more: how in practice the browser extension developer will returnthe address or the balance of a wallet ? Sign a transaction ? Can you propose some pseudo code explaining that ?

Reviewer 3: If I were to request the balance of a wallet inside the code, I would do:

```typescript
// if the window object has the massa object and a method to get all wallet providers
if (window.massa && window.massa.listWalletProviders) {
    // get all wallet providers
    const walletProviders = window.massa.listWalletProviders();
    // choose which provider to use
    const provider = walletProviders[0]; // get some provider based on the metadata e.g.
    // get wallet from that provider
    const wallet = provider.getWallet();
    // use the wallet methods ...
    wallet.getAddress(....)
}
```
As already explained, the Provider itself will communicate with the extension over message dispatching using a controller and `window.postMessage` method. Once the message is posted, the wallet plugin (e.g. a chrome plugin) will have a listener attached listening for messages of a given type and payload and if the message is designated for the plugin (based on the metadata), the plugin will respond to it and send the result back. I will present a schema explaining all of that.

Reviewer: @BatiGencho This code is from the Dapps dev point of view.

Can you put yourself in the shoes of the extension developer and write the very same code to handle "getWallets" and "getAddress" ?

can you write the pseudo code doing: 

>have a listener attached listening for messages of a given type and payload and if the message is designated for the plugin (based on the metadata), the plugin will respond to it and send the result back.

and the one doing:

> Provider itself will communicate with the extension over message dispatching using a controller and window.postMessage method

by considering my previous comment (about events on `window` object) ?

Author: when registered, the bearby wallet will emit an event with the eventTarget name, so that the massa library can create a wallet object. So in what I am proposing, the wallet object is not defined in bearby side but in massa library side. The wallet object is a class with a sign method. (see bellow)

**Code snippet of massa javascript (or typescript) library:**

```text
registeredProviders = []

// we start by listening register event
window.massaWalletProvider.addEventListener('register', (payload) => {
  registeredProviders.push(payload)
})

window.massaWalletProvider.emit('loaded') // all wallet will catch this event and emit register event
```

**As a dapp builder, I would write...**

```text
providers Provider[] = await massa-library.listWalletProviders()

for each providers as p
  for each p.listWallets as wallet
    print wallet.getAddress()
    print wallet.getBalance()

wallet = providers.listWallets()[0]
pkey, signature = await wallet.sign([...])
```

**Code snippet of massa javascript (or typescript) library:**

```text
export function listWalletProviders() {...}

export class Wallet {
  getAddress() {...}
  getBalance() {...}
  sign(payload: bytes): (string, bytes) {...}
}

> export class Provider {

Reviewer: One need is mising from https://github.com/massalabs/massa-standards/issues/13 :

```
a way to manage the wallet if authorized by the user (add/remove keys)
```

Reviewer 3: Here I would refer to what Greg is saying in terms of missing funcs: https://github.com/massalabs/massa-standards/pull/26/files#diff-dad455c0814fa2229eaae244a61b26f2f312670c0d14b6da452cd46744739956R137-R175
Above are some more examples of what we could/must have a wallet expose.

Reviewer: No I don't think that all the functions listed there are needed now. Can you extract the meaningful ones and complete this pseudo code section ?

Reviewer 3: ```typescript
export class Provider {
  // shows the connected network
  getConnectedNetwork(): MassaNetwork;
  // connects to the provider
  async connect(massaNetwork: MassaNetwork);
  // disconnects from the provider
  async disconnect();
  // lists all wallets
  listWallets() -> [Wallet];
}
```

Reviewer: @BatiGencho i don't see how the proposed methods can help covering the need `a way to manage the wallet if authorized by the user (add/remove keys)`. I think what @damir is expecting here is a way to import or export key pairs.

@damip can you confirm?
@SlnPons do you have such need on your side ?

Reviewer: Ok, so the needs is confirm for two more functions:
- importWallet
- deleteWallet

@BatiGencho I will let you add those two functions with the proper signature inthe markdown file.

Author: I simply added importWallet() {...}
  deleteWallet() {...}

just bellow but Provider class is not defined yet.

  listWallets() {...}

>   sayHello() {...}

Reviewer: To be remove

Reviewer 3: agree, see comment above.

  window[`massaWalletProvider-${payload.eventTarget}`] = extensionEventTarget;
  registeredProviders[payload.providerName] = payload.eventTarget;
})

interface Message {
  command: string;
  params: object;
  requestId: string;
}

const pendingRequests = new Map<string, ?>();

//send a message from the webpage script to the content script
function sendMessageToContentScript(provider, command, params, responseCallback) {
  const requestId = uuidv4();
  const message: Message = { command, params, requestId };
  pendingRequests.set(requestId, responseCallback);

  window[`massaWalletProvider-${registeredProviders[provider]}`].dispatchEvent(new CustomEvent('message', { detail: message }));
}

//receive a response from the content script
function handleResponseFromContentScript(event) {
  const { result, error, requestId } = event.detail;
  const responseCallback = pendingRequests.get(requestId);

  if (responseCallback) {
    if (error) {
      responseCallback(new Error(error));
    } else {
      responseCallback(null, result);
    }
    pendingRequests.delete(requestId);
  }
}

window.massaWalletProvider.addEventListener('message', handleResponseFromContentScript);
```

> Because it was requested, here a potential implementation of the `registerAsMassaWalletProvider` in the content script:

Author: I still don't understand how the callback given to the sign function will be executed.

Author: pseudo code implementation in commit fe31ed329ea6400e25cae9d3e687921fe7f996d5

## Specification

## Implementation

## Code sample

This section provides examples of how to use the Massa wallet-provider JS library from different perspectives.
We'll consider two potential users: Alice and Bob.

### Bob, the dApp developer

Bob wants to create a web page that can interact with any wallet provider within the Massa ecosystem.

Bob needs the ability to perform the following tasks:

- Determine if a wallet provider is available.
- Access the names of available wallet providers to personalize website messaging.

> - Retrieve a list of all the wallets registered with each provider.

Reviewer 5: I guess this should be `accounts` 
```suggestion
- Retrieve a list of all the accounts registered with each provider.
```

- Retrieve wallet information, such as address and balance.

> - Perform actions on a wallet, such as signing a transaction or exporting a wallet.

Reviewer 5: ```suggestion
- Perform actions on an account, such as signing a transaction or exporting an account.
```

> - Remove a wallet from a provider.

Reviewer 5: ```suggestion
- Remove a account from a provider.
```

To do this, he can use the Massa wallet-provider JS library:

```typescript
import { providers, Provider, Account, SignResult } from 'massa-wallet-provider';

async function interactWithMassaWallet() {
  // Get the available wallet providers
  const availableProviders: Provider[] = await providers();

  // Display the names of the available wallet providers on the website
  const providerNames: string[] = availableProviders.map(provider => provider.name());
  document.getElementById('provider-names').innerText = providerNames.join(', ');

  // Allow the user to select a wallet provider
  const selectedProvider: Provider = availableProviders[0];

  // Display the accounts registered with the selected provider on the website
  const accounts: Account[] = await selectedProvider.accounts();

>   const walletAddresses: string[] = await Promise.all(accounts.map(async account => await account.address()));
>   document.getElementById('account-addresses').innerText = walletAddresses.join(', ');

Reviewer 5: ```suggestion
  const accountAddresses: string[] = await Promise.all(accounts.map(async account => await account.address()));
  document.getElementById('account-addresses').innerText = accountAddresses.join(', ');
```

  async balance(): Promise<BigInt> { }

  async sign(payload: Uint8Array): Promise<[string, Uint8Array]> { }
}

export class Provider {
  name(): string { }

  async wallets(): Promise<Wallet[]> { }

  async importWallet(): Promise<Wallet> { }

  async deleteWallet(wallet: Wallet): Promise<void> { }
}
```

### Alice, the browser extension developer

Alice wants to create a browser extension that can interact with the Massa ecosystem using the wallet-provider JS library.
However, Alice needs to interact with the library from the [content script](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts), which is separate from the webpage script.

Alice needs the ability to perform the following tasks:

- Register her wallet extension as a provider with the web page.
- Receive commands from the web page and respond to them accordingly.

To do this, she can use the Massa wallet-provider-content-script JS library:

```typescript
import {
  registerAsMassaWalletProvider,
  listWallet,
  deleteWallet,
  importWallet,
  balance,
  sign,
} from 'wallet-provider-content-script';

// Register Alice's wallet extension as a provider with the web page

> registerAsMassaWalletProvider('Alice Wallet').then((success) => {

Author: I don't understand how this function will save this callback and trigger it again when ask by the dapp builder

Reviewer: See [ehre](https://github.com/massalabs/massa-standards/pull/27/commits/29d50a06b196fd93680af1e98d00c700346c5e51#diff-dad455c0814fa2229eaae244a61b26f2f312670c0d14b6da452cd46744739956R247-R260).

});

importWallet((pubKey, privKey) => {
  // Handle importWallet command from the web page
  console.log(`Importing wallet with public key ${pubKey} and private key ${privKey}`);
});

balance((address) => {
  // Handle balance command from the web page
  const balanceAmount = BigInt(1000000);
  return balanceAmount;
});

sign((address, payload) => {
  // Handle sign command from the web page
  const signature = '0x123456...';
  return [signature, payload];
});
```

Alice can utilize our wallet-provider-content-script JS library, which provides the following functions:

```typescript

export async function registerAsMassaWalletProvider(providerName: string): Promise<boolean> {}

export function listWallet(callback: () => string[]): void {}

export function deleteWallet(callback: (address: string) => void): void {}

export function importWallet(callback: (pubKey: string, privKey: string) => void): void {}

export function balance(callback: (address: string) => BigInt): void {}

export function sign(callback: (address: string, payload: Uint8Array) => [string, Uint8Array]): void {}

```

### Constraints to solve by the provided lib

> The library needs to provide a secure and efficient way to communicate between the page script and the content script, which allows for multiple requests to be processed in parallel.

Reviewer: Done [here](https://github.com/massalabs/massa-standards/pull/27/commits/29d50a06b196fd93680af1e98d00c700346c5e51#diff-dad455c0814fa2229eaae244a61b26f2f312670c0d14b6da452cd46744739956R181-R183).

## Specification

## Implementation

## Code sample

This section provides examples of how to use the Massa wallet-provider JS library from different perspectives. We'll consider two potential users: Alice and Bob.

### Bob, the dApp developer
Bob wants to create a web page that can interact with any wallet provider within the Massa ecosystem. 

Bob needs the ability to perform the following tasks:

- Determine if a wallet provider is available.
- Access the names of available wallet providers to personalize website messaging.
- Retrieve a list of all the wallets registered with each provider.
- Retrieve wallet information, such as address and balance.
- Perform actions on a wallet, such as signing a transaction or exporting a wallet.
- Remove a wallet from a provider.

To do this, he can use the Massa wallet-provider JS library: 

```typescript
import { providers, Provider, Wallet } from 'massa-wallet-provider';

async function interactWithMassaWallet() {
  // Get the available wallet providers
  const availableProviders: Provider[] = await providers();

  // Display the names of the available wallet providers on the website

>   const providerNames: string[] = availableProviders.map(provider => provider.name());

Reviewer: Yes, as the providers need to register to the web page script, the list of providers is something known locally, without needed to ask the content script or background script.

So I removed the promise.

  document.getElementById('provider-names').innerText = providerNames.join(', ');

  // Allow the user to select a wallet provider
  const selectedProvider: Provider = availableProviders[0];

  // Display the wallets registered with the selected provider on the website
  const wallets: Wallet[] = await selectedProvider.wallets();

>   const walletAddresses: string[] = await Promise.all(wallets.map(async wallet => wallet.address()));

Reviewer: :man_shrugging:

  document.getElementById('wallet-addresses').innerText = walletAddresses.join(', ');

  // Allow the user to select a wallet
  const selectedWallet: Wallet = wallets[0];

  // Display the balance of the selected wallet on the website
  const walletBalance: BigInt = await selectedWallet.balance();
  document.getElementById('wallet-balance').innerText = walletBalance.toString();

  // Allow the user to sign a transaction
  const payload: Uint8Array = new Uint8Array([1, 2, 3]);
  const [pubKey, signature]: [string, Uint8Array] = await selectedWallet.sign(payload);
}
```

Bob can utilize our wallet-provider JS library, which provides the following functions:

```typescript
export async function providers(): Promise<Provider[]> { }

export class Wallet {
  async address(): Promise<string> { }

>   async balance(): Promise<BigInt> { }

Reviewer 4: It looks like you use the wallet implementation of Thyra but the one in massa-client is different. Thyra has only one account wallet and massa-client has multiple account per wallet. https://github.com/massalabs/massa/blob/a3c6e0373c9dab8bd7bc0f17686d3de1561d54d6/massa-wallet/src/lib.rs#L25

Should we create a wallet standard in the same way of the unit standard to be able to define both dapp-wallet communication and wallet file standards ?

Reviewer: > Should we create a wallet standard in the same way of the unit standard to be able to define both dapp-wallet communication and wallet file standards ?

I'm not sure. You have pretty much solved this vocabulary problem in your comment by using `Account`.

My suggestion is to use the concept of `account` instead of `wallet` :
- on the thyra side when we talk about the key/address pair wrapper
- here by renaming the `Wallet` class to `Account`.

@Ben-Rey is this an appropriate response to your comment?

@SlnPons is this something acceptable to the end user in this context? on the thyra side?

Reviewer 4: Perfect that fixes the problem.

Reviewer 6: @Ben-Rey @gregLibert yep, it matches thyra wallet coming changes. Where we'll talk about accounts! Al good on my side.

Author: on Thyra: 1 account = 1 wallet

on massa-client: one wallet can contains multiple account

Wallet = thyra plugin wallet | browser

Account = 1 key pair

  // Allow the user to select a wallet provider
  const selectedProvider: Provider = availableProviders[0];

  // Display the accounts registered with the selected provider on the website
  const accounts: Account[] = await selectedProvider.accounts();
  const walletAddresses: string[] = await Promise.all(accounts.map(async account => await account.address()));
  document.getElementById('account-addresses').innerText = walletAddresses.join(', ');

  // Allow the user to select an account

>   const selectedWallet: Account = accounts[0];

Reviewer 4: selectedAccount

  // Display the balance of the selected account on the website

>   const walletBalance: BigInt = await selectedWallet.balance();

Reviewer 4: accountBalance
